feat: support for custom templates

hacdias opened the pull request:
Closes #125 (see for more context). This PR:

- adds `-templates` flag that can be used to provide a directory with templates. They override existing ones for users that want to customize how the pages look.
- added a template for 404 pages
- compressed template handling into `serveHTML`

hacdias commented:
@abhinav just pinging you, in case you have a chance to look at this 😄 Also, thanks for merging the other PRs!

abhinav commented on handler.go:
```diff
@@ -2,21 +2,19 @@ package main
+	//go:embed templates/*.html
+	templateFiles embed.FS
+
+	templates = template.Must(template.ParseFS(templateFiles, "templates/*.html"))
```
Minor: code here doesn't always comply with this, but with a broad name like "templates", we should add a `_` prefix to the unexported global -- since there's a good chance there will be a function with a local variable named 'templates'.

ref: https://github.com/uber-go/guide/blob/master/style.md#prefix-unexported-globals-with-_

abhinav commented on main.go:
```diff
@@ -20,6 +22,14 @@ func main() {
 		log.Fatalf("Failed to parse %s: %v", *yml, err)
 	}
 
+	if *tpls != "" {
+		log.Printf("Parsing templates at path: %s\n", *tpls)
+		templates, err = templates.ParseGlob(filepath.Join(*tpls, "*.html"))
+		if err != nil {
```
This is good, but I'd prefer if we didn't overwrite the global variable.

Could we pick between `_templates`and the parsed custom templates here in main and pass that to `CreateHandler`, and plumb it down to the individual handlers?

abhinav reviewed (commented):
Thanks for this and the other PRs, @hacdias, and thanks for waiting. I had to make some time to look over this. (The other PRs were a lot quicker to review. 😅)

I'm in favor of this change in general, and I appreciate the rigor that went into this. This is a good PR. I have only a couple minor suggestions. It should be good to merge after that.

Thanks!

hacdias commented on handler.go:
```diff
@@ -32,7 +31,22 @@ var (
 //		Package page for the given subpackage.
-func CreateHandler(config *Config) http.Handler {
+func CreateHandler(config *Config, templates *template.Template) (http.Handler, error) {
+	indexTemplate := templates.Lookup("index.html")
+	if indexTemplate == nil {
+		return nil, errors.New("template index.html is missing")
+	}
+
+	notFoundTemplate := templates.Lookup("404.html")
+	if indexTemplate == nil {
+		return nil, errors.New("template 404.html is missing")
+	}
+
+	packageTemplate := templates.Lookup("package.html")
+	if indexTemplate == nil {
+		return nil, errors.New("template package.html is missing")
+	}
```
ℹ️ This errors should **never** happen, as the default templates will always be there.

hacdias commented on handler_test.go:
```diff
@@ -297,6 +336,6 @@ func BenchmarkHandlerDispatch(b *testing.B) {
 type nopResponseWriter struct{}
 
-func (nopResponseWriter) Header() http.Header       { return nil }
+func (nopResponseWriter) Header() http.Header       { return http.Header{} }
```
ℹ️ We now set headers. Benchmark would panic without this.

abhinav replied:
The comment is helpful; thanks!

hacdias commented on handler_test.go:
```diff
@@ -271,7 +309,8 @@ func BenchmarkHandlerDispatch(b *testing.B) {
 			},
 		},
-	})
+	}, template.Must(_templates.Clone()))
```
ℹ️ Cannot use `getTestTemplates` here because `testing.T != testing.B`.

abhinav replied:
I think `getTestTemplates` can be made to accept a `testing.TB` interface, in which case it'll accept both. 
(Unless there's something testing.T-specific being used in it that I missed.)

hacdias commented on utils_test.go:
```diff
@@ -17,43 +19,39 @@ import (
-	if _, err := tmpfile.Write(content); err != nil {
-		t.Fatal("Unable to write tmpfile", err)
-	}
-	if err := tmpfile.Close(); err != nil {
-		t.Fatal("Unable to close tmpfile", err)
-	}
+	_, err = tmpfile.Write(content)
+	require.NoError(t, err, "unable to write tmpfile")
+
+	err = tmpfile.Close()
+	require.NoError(t, err, "unable to close tmpfile")
```
ℹ️ Here and bellow just some cleanup to match what is done in other tests.

hacdias commented:
@abhinav thanks! I addressed your feedback, and also cleaned up a few things I saw along the way. Looking forward to see this merged. I found many packages like this one, but none did _exactly_ what I wanted.

abhinav commented on config_test.go:
```diff
@@ -9,7 +9,7 @@ import (
 
 func TestParse(t *testing.T) {
-	path, clean := TempFile(t, `
+	path := TempFile(t, `
```
Nice, thanks for t.Cleanup-ing this.

abhinav commented on handler.go:
```diff
@@ -32,7 +32,22 @@ var (
-func CreateHandler(config *Config) http.Handler {
+func CreateHandler(config *Config, templates *template.Template) (http.Handler, error) {
+	indexTemplate := templates.Lookup("index.html")
+	if indexTemplate == nil {
+		return nil, errors.New("template index.html is missing")
+	}
+
+	notFoundTemplate := templates.Lookup("404.html")
+	if notFoundTemplate == nil {
+		return nil, errors.New("template 404.html is missing")
+	}
+
+	packageTemplate := templates.Lookup("package.html")
+	if packageTemplate == nil {
+		return nil, errors.New("template package.html is missing")
+	}
```
👌 
appreciate resolving the templates here.

abhinav commented on handler.go:
```diff
@@ -169,24 +186,38 @@ func (h *packageHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 }
+
+func serveHTML(w http.ResponseWriter, status int, template *template.Template, data interface{}) {
+	if status >= 400 {
+		w.Header().Set("Cache-Control", "no-cache")
+	}
```
Resolves #18

Thanks!

abhinav commented on handler_test.go:
```diff
@@ -260,8 +273,56 @@ func TestIndexHandler_rangeOf(t *testing.T) {
 }
 
+func TestCustomTemplates(t *testing.T) {
+	t.Run("fails to create handler if template is missing", func(t *testing.T) {
```
minor: can use a shorter name here ("custom templates" is already in context)

```suggestion
	t.Run("missing", func(t *testing.T) {
```

abhinav commented on handler_test.go:
```diff
@@ -260,8 +273,56 @@ func TestIndexHandler_rangeOf(t *testing.T) {
+			"404.html": "not found: {{ .Path }}",
+		})
+
+		// Overrides 404.html
+		rr := CallAndRecord(t, config, templates, "/blah")
+		require.Equal(t, http.StatusNotFound, rr.Result().StatusCode)
+
+		// But not package.html
```
👍

abhinav commented on handler_test.go:
```diff
@@ -260,8 +273,56 @@ func TestIndexHandler_rangeOf(t *testing.T) {
+		}
+	})
+
+	t.Run("only replaces correct templates", func(t *testing.T) {
```
```suggestion
	t.Run("replace", func(t *testing.T) {
```

abhinav commented on main.go:
```diff
@@ -20,9 +23,35 @@ func main() {
+	if err != nil {
+		return nil, err
+	}
+	templates, err = templates.ParseGlob(filepath.Join(dir, "*.html"))
+	return templates, err
```
optional: can shorten to

```suggestion
	return templates.ParseGlob(filepath.Join(dir, "*.html"))
```

abhinav commented on utils_test.go:
```diff
@@ -67,6 +66,29 @@ func AssertResponse(t *testing.T, rr *httptest.ResponseRecorder, code int, want
+// getTestTemplates returns a [template.Template] object with the default templates,
+// overwritten by the  [overrideTemplates]. If [overrideTemplates] is nil, the returned
+// templates are a clone of the global [_templates].
+func getTestTemplates(t *testing.T, overrideTemplates map[string]string) *template.Template {
```
Yeah, I think this can just use `testing.TB`.
```suggestion
func getTestTemplates(t testing.TB, overrideTemplates map[string]string) *template.Template {
```

abhinav reviewed (approved):
No blockers. This is great! Thanks @hacdias!

I'm okay with this as-is, but if you'd like to make `getTestTemplates` take a `testing.TB`, I can wait a day.

hacdias commented:
@abhinav thanks for the review! I applied the feedback, and I think it's ready to be merged.

abhinav commented:
Merged! Thanks, @hacdias
